### PR TITLE
refactor daemon image tagging

### PR DESCRIPTION
**Pull request by MHBauer**
Using this PR to start discussion.

daemon exposes directly it's internal `repositories` member, which is a `*graph.TagStore`. Any external caller then needs to be able to import the docker/graph package. Most of these functions return only an error. The docker/image/Image struct will need to be moved into the api at some point in the future. graph.ImagePullConfig needs to move as well.

If this looks okay in principle, I would then do something similar for all of the remaining `daemon.Repositories().XYZ` calls.

Here is a list of what remains:

```
λ ag '\.Repositories\(\)'
api/server/image.go
115:        err = s.daemon.Repositories().Pull(ctx, image, tag, imagePullConfig)
132:        err = s.daemon.Repositories().Import(ctx, src, repo, tag, message, r.Body, output, newConfig)
187:    if err := s.daemon.Repositories().Push(ctx, name, imagePushConfig); err != nil {
215:    if err := s.daemon.Repositories().ImageExport(names, output); err != nil {
226:    return s.daemon.Repositories().Load(r.Body, w)
259:    imageInspect, err := s.daemon.Repositories().Lookup(vars["name"])
367:    images, err := s.daemon.Repositories().Images(r.Form.Get("filters"), r.Form.Get("filter"), boolValue(r, "all"))
381:    history, err := s.daemon.Repositories().History(name)

builder/dispatchers.go
212:    image, err := b.Daemon.Repositories().LookupImage(name)

builder/internals.go
519:    if err := b.Daemon.Repositories().Pull(ctx, remote, tag, imagePullConfig); err != nil {
523:    image, err := b.Daemon.Repositories().LookupImage(name)

daemon/image_delete.go
57: img, err := daemon.Repositories().LookupImage(imageRef)
94:     repoRefs := daemon.Repositories().ByID()[img.ID]
120:    return len(daemon.Repositories().ByID()[imageID]) > 1
149:    _, err := daemon.Repositories().Delete(repository, ref)
160:    imageRefs := daemon.Repositories().ByID()[imgID]
312:    if daemon.Repositories().HasReferences(img) {
341:    return !(daemon.Repositories().HasReferences(img) || daemon.Graph().HasChildren(img))

daemon/list.go
167:            image, err := daemon.Repositories().LookupImage(ancestor)
296:    img, err := daemon.Repositories().LookupImage(container.Config.Image)
```

### CODE REVIEW

**Comment by LK4D4** [comment]
Idea sounds good to me. But why you need to be able to import `docker/graph`?

**Comment by MHBauer** [comment]
We want to *avoid* importing docker/graph, or anything else in base docker. At which point we can rip the API out of where it is and into a pkg/.

**Comment by MHBauer** [comment]
I really need to stop forgetting to @calavera @icecrime @duglin

**Comment by cpuguy83** [comment]
Seems good, but I'd call it just `PullImage` and `TagImage` unless there is some conflict there.

**Comment by duglin** [comment]
Yep, looks right. Agree with @cpuguy83 on the shorter names, if possible.  Needs a rebase.

**Comment by MHBauer** [comment]
Thanks for the tips. I'll continue on with shorter names.

**Comment by MHBauer** [comment]
As I get down to the end, should I continue to remove all external references to daemon.Repositories() and then privatize it? Or do only the work necessary to cut the tie in api/ ?

**Comment by LK4D4** [comment]
@MHBauer makes sense for consistency to do that. In other PR of course.

**Comment by LK4D4** [comment]
I changing to code-review, because seems like we agree on this and it more like refactoring, than design decision.

**Comment by LK4D4** [comment]
@MHBauer golint failure

File: daemon/daemon.go [hunk 1]
```diff
@@ -984,6 +985,62 @@ func (daemon *Daemon) Repositories() *graph.TagStore {
...
+	return daemon.repositories.Push(localName, imagePushConfig)
+}
+
+// LookupImage looks up an image by name and returns it as an ImageInspect
+// structure.
+func (daemon *Daemon) LookupImage(name string) (*types.ImageInspect, error) {
+	return daemon.repositories.Lookup(name)
+}
+
+// Load uploads a set of images into the repository. This is the complementary of ImageExport.
```
Comment (vdemeester) [inline daemon/daemon.go]: s/Load/LoadImage/

File: daemon/daemon.go [hunk 2]
```diff
@@ -984,6 +985,62 @@ func (daemon *Daemon) Repositories() *graph.TagStore {
...
+func (daemon *Daemon) LoadImage(inTar io.ReadCloser, outStream io.Writer) error {
+	return daemon.repositories.Load(inTar, outStream)
+}
+
+// Images returns a filtered list of images. filterArgs is a JSON-encoded set
+// of filter arguments which will be interpreted by pkg/parsers/filters.
+// filter is a shell glob string applied to repository names. The argument
+// named all controls whether all images in the graph are filtered, or just
+// the heads.
+func (daemon *Daemon) ListImages(filterArgs, filter string, all bool) ([]*types.Image, error) {
```
Comment (vdemeester) [inline daemon/daemon.go]: s/Images/ListImages/

**Comment by vdemeester** [comment]
Looks good ! @MHBauer I pointed out the 2 golint failures :stuck_out_tongue_closed_eyes:

**Comment by MHBauer** [comment]
This takes care of everything necessary to isolate api/server. I'm having trouble running the validate section of the makefile, but the tests pass. I'll wait for CI to do a build and then squash.

File: daemon/daemon.go [hunk 3]
```diff
@@ -984,6 +985,68 @@ func (daemon *Daemon) Repositories() *graph.TagStore {
...
+	return daemon.repositories.Images(filter, filter, all)
+}
+
+// History returns a slice of ImageHistory structures for the specified image
```
Comment (vdemeester) [inline daemon/daemon.go]: s/History/ImageHistory/

**Comment by vdemeester** [comment]
@MHBauer what problem do you have with the `validate` section ? :sweat_smile: (I know this is not really related to the PR, but I'm curious)

**Comment by MHBauer** [comment]
@vdemeester I'm seeing this:
```
Step 46 : COPY . /go/src/github.com/docker/docker
 ---> a1310fe013f4
Removing intermediate container 76ea10924de6
Successfully built a1310fe013f4
docker run --rm -it --privileged -e BUILDFLAGS -e DOCKER_CLIENTONLY -e DOCKER_DEBUG -e DOCKER_EXECDRIVER -e DOCKER_EXPERIMENTAL -e DOCKER_GRAPHDRIVER -e DOCKER_STORAGE_OPTS -e DOCKER_USERLANDPROXY -e TESTDIRS -e TESTFLAGS -e TIMEOUT  "docker-dev:daemon-derepo" hack/make.sh validate-dco validate-gofmt validate-pkg validate-lint validate-test validate-toml validate-vet

---> Making bundle: validate-dco (in bundles/1.9.0-dev/validate-dco)
fatal: unable to access 'https://github.com/docker/docker.git/': Could not resolve host: github.com
make: *** [validate] Error 128
```

I probably just need to restart the VM I am using.

**Comment by duglin** [comment]
failure looks unrelated.
LGTM

ping @jfrazelle @vdemeester

**Comment by jfrazelle** [comment]
LGTM

**Pull request merged**